### PR TITLE
ara::core::Byte implementation 

### PR DESCRIPTION
**Pull request by mwiton**
Implementing requirements for ara::core::Byte type (SWS_CORE_10101 - SWS_CORE_10110). Created wrapper class for std::byte and defining all functions for ara::core::byte, as for std::byte in the standard library.

Issue: #41

### CODE REVIEW

File: include/ara/core/byte.h [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+ */
+
+#ifndef ARA_CORE_BYTE_H_
+#define ARA_CORE_BYTE_H_
+
+#include <cstddef>
+
+namespace ara::core {
+/**
+ * @brief Wrapper class for std::byte class object
```
Comment (bburda) [inline include/ara/core/byte.h:15]: Please end each `@brief` with a `.`, this should be a sentence, so Doxygen knows when to finish parsing.

File: include/ara/core/byte.h [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+/**
+ * @brief Gives value of Byte object as integer.
+ *
+ * @tparam IntegerType Exact type of integer to return
+ * @param b Byte object
+ * @return constexpr IntegerType Integer value
+ */
+template<class IntegerType> constexpr IntegerType
+to_integer(ByteImpl b) noexcept requires std::integral<IntegerType>
```
Comment (bburda) [inline include/ara/core/byte.h:83]: That's great that you used concept here, but in my opinion, in API headers, we should strive to have the C++11 compliant code. Could you use `enable_if` instead?

File: include/ara/core/byte.h (hunk 1 above)
Comment (mwiton) [reply include/ara/core/byte.h]: Added `.` on the end of each `@brief`.

File: include/ara/core/byte.h (hunk 2 above)
Comment (mwiton) [reply include/ara/core/byte.h]: Concepts are changed, instead used `std::enable_if<std::is_integral<IntegerType>::value>::type` for checking if integer.

File: include/ara/core/byte.h [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+
+    /**
+     * @brief Construct a new Byte Impl object with no defined value.
+     *
+     * @req {SWS_CORE_10104}
+     */
+    ByteImpl() = default;
+    /**
+     * @brief Constuct object with given value.
+     * @arg Value of byte
```
Comment (bburda) [inline include/ara/core/byte.h:32]: Should be `@param` right?

File: include/ara/core/byte.h [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+
+#include <cstddef>
+
+namespace ara::core {
+/**
+ * @brief Wrapper class for std::byte class object.
+ *
+ * @req {SWS_CORE_10100}
+ */
+class ByteImpl
```
Comment (bburda) [inline include/ara/core/byte.h:19]: Maybe simple `Byte` without aliasing in `utility`?

File: include/ara/core/byte.h [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+    }
+
+    /**
+     * @brief Equality operator for Byte object.
+     *
+     * @return true When given Byte object is equal
+     * @return false When given Byte object is not equal
+     * @req {SWS_CORE_10109}
+     */
+    constexpr bool operator==(const ByteImpl& secondByteObject) const
```
Comment (bburda) [inline include/ara/core/byte.h:64]: Missing `@param` and you can use `rhs` name for this param.

File: include/ara/core/byte.h [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+    }
+
+    /**
+     * @brief Not-equality operator for Byte object.
+     *
+     * @return true When given Byte object is not equal
+     * @return false When given Byte object equal
+     * @req {SWS_CORE_10110}
+     */
+    constexpr bool operator!=(const ByteImpl& secondByteObject) const
```
Comment (bburda) [inline include/ara/core/byte.h:76]: Same as in `operator==`

File: include/ara/core/utility.h [hunk 7]
```diff
@@ -9,14 +9,16 @@
 
 #include <utility>
 
+#include "byte.h"
```
Comment (bburda) [inline include/ara/core/utility.h:12]: Shall be `"ara/core/byte.h"`

File: include/ara/core/byte.h (hunk 3 above)
Comment (mwiton) [reply include/ara/core/byte.h]: Changed to param.

File: include/ara/core/byte.h (hunk 5 above)
Comment (mwiton) [reply include/ara/core/byte.h]: Done

File: include/ara/core/byte.h (hunk 6 above)
Comment (mwiton) [reply include/ara/core/byte.h]: Changed as in `operator==`

File: include/ara/core/utility.h (hunk 7 above)
Comment (mwiton) [reply include/ara/core/utility.h]: Done

File: include/ara/core/byte.h (hunk 4 above)
Comment (mwiton) [reply include/ara/core/byte.h:19]: There is functional requirement in AUTOSAR specification:
obraz
It says that Byte should be an alias for implementation defined type (in our case distinct class). We implement all requirements for API as are given in documentation or we can change a little?

**Pull request merged**